lib/src/model/nutriscore.dart

```diff
@@ -0,0 +1,208 @@
+      throw FormatException(
+          'Nutri-Score data is not Map<String, dynamic>, got ${json.runtimeType}');
+    }
+
+    final result = <NutriScoreVersion, NutriScore>{};
+
+    for (final MapEntry(:key, :value) in json.entries) {
+      final version = NutriScoreVersion.tryParse(key);
+      if (version == null || value is! Map<String, dynamic>) {
+        continue;
+      }
+
+      try {
+        final raw = NutriScoreRaw.fromJson(value);
+        result[version] = NutriScore._fromRaw(version, raw);
+      } catch (_) {
+        continue;
+      }
+    }
+    return result.isEmpty ? null : result;
+  }
+
+  /// Parses the raw grade string from the API into a [NutriScoreStatus] and optional [NutriScoreGrade].
+  ///
+  /// Returns a tuple where the status indicates whether the grade is computed, unknown, or not applicable.
+  static (NutriScoreStatus, NutriScoreGrade?) _parseGrade(String grade) =>
+      switch (grade.toLowerCase()) {
+        'not-applicable' => (NutriScoreStatus.notApplicable, null),
+        'a' => (NutriScoreStatus.computed, NutriScoreGrade.A),
+        'b' => (NutriScoreStatus.computed, NutriScoreGrade.B),
+        'c' => (NutriScoreStatus.computed, NutriScoreGrade.C),
+        'd' => (NutriScoreStatus.computed, NutriScoreGrade.D),
+        'e' => (NutriScoreStatus.computed, NutriScoreGrade.E),
+        _ => (NutriScoreStatus.unknown, null),
+      };
+
+  /// Parses a [NutriScore] from raw API data using the given [version].
+  ///
+  /// Internally determines the Nutri-Score status, grade, category, and score.
+  /// Falls back to [NutriScore.invalid] if critical fields are missing.
+  static NutriScore _fromRaw(
+    NutriScoreVersion version,
+    NutriScoreRaw raw,
+  ) {
+    final (status, grade) = _parseGrade(raw.grade ?? '');
+    final category = NutriScoreCategory.fromRaw(version, raw.data);
+    final score = raw.score;
+    final missingData = <NutriScoreInput>[
+      if (!raw.categoryAvailable) NutriScoreInput.category,
+      if (!raw.nutrientsAvailable) NutriScoreInput.nutrients,
+    ];
+
+    return switch (status) {
+      NutriScoreStatus.unknown =>
+        NutriScore.unknown(version, category, missingData),
+      NutriScoreStatus.notApplicable => NutriScore.notApplicable(
+          version, category, missingData, raw.notApplicableCategory),
+      NutriScoreStatus.computed
+          when grade == null || category == null || score == null =>
+        NutriScore.invalid('Missing required fields (grade, category, score)'),
+      NutriScoreStatus.computed =>
+        NutriScore.computed(version, grade!, category!, score!),
+      _ => NutriScore.invalid('Unexpected NutriScore: ${raw.grade}'),
+    };
+  }
+
+  @override
+  String toString() {
+    return 'NutriScore{status: ${status.name}, grade: ${grade?.name}, version: $version, category: $category, score: $score, notApplicableCategory: $notApplicableCategory, errorMessage: $errorMessage, missingData: $missingData}';
```

Comment:
I'm sure we can afford some line jumps like that:
```suggestion
    return 'NutriScore{status: ${status.name}'
      ', grade: ${grade?.name}'
      ', version: $version'
      ', category: $category'
```

Reply:
@monsieurtanuki , I intentionally left out `toJson` because the Nutri-Score details are computed on the server side and thus `nutriscore` should be a read-only field and not sent back to the server. Or is there a technical reason to always require `toJson`? In that case I would rather implement a dummy `toJson` that provides `null` or an empty object. What do you think?

Regarding server result tests and formatting: sure, I can add this.

Reply:
You're right, we don't send back this data to the server.
But, as I explained in another comment
> We need a `toJson` too, as in Smoothie we spend our time storing products as json strings in the local database and retrieving them when relevant

I meant a meaningful `toJson`.
Actually in Smoothie we never display data from the server. We get data from the server and then store it in our local database. Then, we extract data from the local database to display it.
For that we use both fromJson and toJson.